GoogleSignIn/Sources/GIDAppCheck/Implementations/GIDAppCheck.m

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright 2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#import "GoogleSignIn/Sources/GIDAppCheck/Implementations/GIDAppCheck.h"
+#import "GoogleSignIn/Sources/GIDAppCheck/API/GIDAppCheckProvider.h"
+#import "GoogleSignIn/Sources/GIDAppCheckTokenFetcher/Implementations/FIRAppCheck+GIDAppCheckTokenFetcher.h"
+
+@import FirebaseAppCheck;
+
+#if TARGET_OS_IOS && !TARGET_OS_MACCATALYST
+
+/// Identifier for queue where App Check work is performed.
+static NSString *const kGIDAppCheckQueue = @"com.google.googlesignin.appCheckWorkerQueue";
+
+@interface GIDAppCheck ()
+
+@property(nonatomic, strong) id<GIDAppCheckTokenFetcher> tokenFetcher;
+@property(nonatomic, strong) dispatch_queue_t workerQueue;
+@property(atomic, getter=isPrepared) BOOL prepared;
+
+@end
+
+@implementation GIDAppCheck
+
+- (instancetype)initWithAppCheckTokenFetcher:(nullable id<GIDAppCheckTokenFetcher>)tokenFetcher {
+  if (self = [super init]) {
+    _prepared = NO;
+    _tokenFetcher = tokenFetcher ?: [FIRAppCheck appCheck];
+    _workerQueue = dispatch_queue_create("com.google.googlesignin.appCheckWorkerQueue", nil);
+  }
+  return self;
+}
+
+- (void)prepareForAppCheckWithCompletion:(nullable void (^)(FIRAppCheckToken * _Nullable,
+                                                            NSError * _Nullable))completion {
```

Comment:
Can we prevent doing the following routine twice if there's a second call coming in before the first call returns?

The `prepared` bit isn't sufficient for that purpose, we'd need a bit like `preparationPending`.

Reply:
I could see how that'd be helpful, but I can also see how that would be confusing. For example, what should we do if `preparing` is true? Do we call completion with an error?

```
- (void)prepareForAppCheckWithCompletion:(nullable void (^)(FIRAppCheckToken * _Nullable,
                                                            NSError * _Nullable))completion {
  if (self.preparing) {
    completion(nil, someErrorSayingThatPreparationIsAlreadyPending);
    return;
  }
  ...
}
```

I think the confusing part is that the caller, potentially at a different location than where the other pending prepare message was sent, will not receive a token or any indication of where to look for confirmation that preparation succeeded.

The way this is currently handled is to not guard against multiple inbound preparation requests, but to instead schedule those requests in order on the same async queue. This strategy doesn't prevent multiple, and wasteful, prepare requests, but it does avoid the confusion I mention above.

What do you think?

Reply:
What you said makes sense. However, from the perspective of protecting the backend service that generates App Check tokens, we should make the API as safe as possible.

For multiple callbacks, a common technique is to save the callbacks in an array and invoked them altogether when the async operation is complete.

I think it's ok to leave a TODO for now and come back to it at a later point.

(Btw, do we need to return the token in the callback of this method?)

Reply:
Fair enough. I've made a first pass at this in https://github.com/google/GoogleSignIn-iOS/pull/314/commits/68510d7fb22d304c306e298686771eb2036c8753. Let me know what you think. Happy to go thru feedback if you think it's the right approach. Otherwise, I can make an issue and put it there to address soon.

Reply:
>  do we need to return the token in the callback of this method?

I'm a little torn. It is a little confusing to return the token here since this is just a prepare method; however, returning the token is a good signal of success. But perhaps we can remove the token parameter from completion and therefore the presence or not of an `NSError` would signal the same thing. What do you think?

Reply:
Given that we don't actually need a token at preparation time, I'd suggest we remove it. I think having just an `NSError` indicating success/failure of the preparation is pretty clear and self-explanatory.

Reply:
SGTM